### ISSUE TITLE
Secret key encryption

### ISSUE THREAD

**Issue by x-logic-iot** [tpm2-software/tpm2-tools#2827]
Hello, I am trying to store some random keys to the TPM. The key should never leave the TPM unencrypted, and when it leaves the TPM encrypted, it is encrypted with a key that only the TPM can derive/know. Any ideas on which commands should I use?

**Comment by williamcroberts** [comment]
So If I understand correctly, you generate keys external to the TPM that you want to load into the TPM and then if they get exported they must leave in an encrypted fashion, if so, then:

To import a key, use `tpm2_import` command.

To export a key, use `tpm2_export` command.

If you set the attribute `fixedparent` in `tpm2_import` via `-a`, along with all the other attributes you need for the key, you can keep the key from being duplicated if you desire.

**Comment by x-logic-iot** [comment]
Thank you for your response. I managed to import the key with tpm2_import, but I can't export it because tpm2_export doesn't exist anymore.

**Comment by idesai** [comment]
The man page example for **[tpm2_duplicate](https://tpm2-tools.readthedocs.io/en/latest/man/tpm2_duplicate.1/)** is a good example.

**Comment by x-logic-iot** [comment]
I have tried following the tpm2_duplicate example, but I get an error after this command:

```
tpm2_create -C primary.ctxt -g sha256 -G rsa -r key.prv -u key.pub \
-L policy.dat -a "sensitivedataorigin"
```

Returns this:

```
WARNING:esys:src/tss2-esys/api/Esys_Create.c:398:Esys_Create_Finish() Received TPM Error
ERROR:esys:src/tss2-esys/api/Esys_Create.c:134:Esys_Create() Esys Finish ErrorCode (0x000002c2)
ERROR: Esys_Create(0x2C2) - tpm:parameter(2):inconsistent attributes
ERROR: Unable to run tpm2_create
```

**Comment by williamcroberts** [comment]
Weird that the manpage shows us manually setting attributes like that. Try dropping the `-a` argument to tpm2_create. The reason it's failing is that non-keyedhash objects, in this case rsa, need to have both encrypt and/or decrypt attribute set.

@idesai is the manpage wrong, should we just drop the `-a` option to `tpm2_create` here?

**Comment by x-logic-iot** [comment]
It seems to me that I have found the solution with `tpm2_rsaencrypt`. First I encrypt the secret with `tpm2_rsaencrypt` and then store it to the TPM register with `tpm2_evictcontrol`.  That way only my TPM can decrypt the secret with `tpm2_rsadecrypt`. Please correct me if I am wrong.

Also, I have tried using [tpm2_encryptdecrypt](https://tpm2-tools.readthedocs.io/en/latest/man/tpm2_encryptdecrypt.1/) and I got an error after this command `tpm2_encryptdecrypt -c key.ctx -o secret.enc secret.dat`

```
WARN: Using a weak IV, try specifying an IV
WARNING:esys:src/tss2-esys/api/Esys_EncryptDecrypt2.c:322:Esys_EncryptDecrypt2_Finish() Received TPM Error
ERROR:esys:src/tss2-esys/api/Esys_EncryptDecrypt2.c:107:Esys_EncryptDecrypt2() Esys Finish ErrorCode (0x000b0143)
WARNING:esys:src/tss2-esys/api/Esys_EncryptDecrypt.c:328:Esys_EncryptDecrypt_Finish() Received TPM Error
ERROR:esys:src/tss2-esys/api/Esys_EncryptDecrypt.c:110:Esys_EncryptDecrypt() Esys Finish ErrorCode (0x000b0143)
ERROR: Esys_EncryptDecrypt(0xB0143) - rmt:error(2.0): command code not supported
ERROR: Unable to run tpm2_encryptdecrypt
```

**Comment by williamcroberts** [comment]
> It seems to me that I have found the solution with `tpm2_rsaencrypt`. First I encrypt the secret with `tpm2_rsaencrypt` and then store it to the TPM persistent memory with `tpm2_evictcontrol`. Then, after unsealing the secret with `tpm2_unseal`, only my TPM can decrypt the secret with `tpm2_rsadecrypt`. Please correct me if I am wrong.

I have no idea what you're trying to say here, can you try again?

> 
> Also, I have tried using [tpm2_encryptdecrypt](https://tpm2-tools.readthedocs.io/en/latest/man/tpm2_encryptdecrypt.1/) and I got an error after this command `tpm2_encryptdecrypt -c key.ctx -o secret.enc secret.dat`
> 
> ```
> WARN: Using a weak IV, try specifying an IV
> WARNING:esys:src/tss2-esys/api/Esys_EncryptDecrypt2.c:322:Esys_EncryptDecrypt2_Finish() Received TPM Error
> ERROR:esys:src/tss2-esys/api/Esys_EncryptDecrypt2.c:107:Esys_EncryptDecrypt2() Esys Finish ErrorCode (0x000b0143)
> WARNING:esys:src/tss2-esys/api/Esys_EncryptDecrypt.c:328:Esys_EncryptDecrypt_Finish() Received TPM Error
> ERROR:esys:src/tss2-esys/api/Esys_EncryptDecrypt.c:110:Esys_EncryptDecrypt() Esys Finish ErrorCode (0x000b0143)
> ERROR: Esys_EncryptDecrypt(0xB0143) - rmt:error(2.0): command code not supported
> ERROR: Unable to run tpm2_encryptdecrypt
> ```

Most TPM's dont support symmetric encryption, so this is fairly common.

**Comment by x-logic-iot** [comment]
Basically, I want to store a secret, (for example "password123") to the TPM persistent memory. When I request that secret it needs to come out only in encrypted form and only TPM should be able to decrypt it. I tried to do it with `tpm2_rsaencrypt` and `tpm2_unseal`, but I got an error after trying to unseal the encrypted secret:

> WARNING:esys:src/tss2-esys/api/Esys_Unseal.c:295:Esys_Unseal_Finish() Received TPM Error
> ERROR:esys:src/tss2-esys/api/Esys_Unseal.c:98:Esys_Unseal() Esys Finish ErrorCode (0x0000018a)
> ERROR: Esys_Unseal(0x18A) - tpm:handle(1):the type of the value is not appropriate for the use
> ERROR: Unable to run tpm2_unseal

You suggested that I should use `tpm2_import` and `tpm2_export`. Since `tpm2_export` doesn't exist anymore, is there any alternative?

**Comment by williamcroberts** [comment]
> Basically, I want to store a secret, (for example "password123") to the TPM persistent memory. When I request that secret it needs to come out only in encrypted form and only TPM should be able to decrypt it. I tried to do it with `tpm2_rsaencrypt` and `tpm2_unseal`, but I got an error after trying to unseal the encrypted secret:
> 
> > WARNING:esys:src/tss2-esys/api/Esys_Unseal.c:295:Esys_Unseal_Finish() Received TPM Error
> > ERROR:esys:src/tss2-esys/api/Esys_Unseal.c:98:Esys_Unseal() Esys Finish ErrorCode (0x0000018a)
> > ERROR: Esys_Unseal(0x18A) - tpm:handle(1):the type of the value is not appropriate for the use
> > ERROR: Unable to run tpm2_unseal
> 
> You suggested that I should use `tpm2_import` and `tpm2_export`. Since `tpm2_export` doesn't exist anymore, is there any alternative?

So in your model you release an encrypted blob from the TPM to your application, then send it back to decrypt it and then send the plaintext from the TPM back to your application. The extra trips for encryption don't make sense since it leaves the TPM at some point in plain-text. tpm2_create with the `-i` option will seal any data, so if you can't unseal it that's an issue in your script. That sealed data is protected by a key only that TPM knows.

For protecting the unsealed secret to your application,  I would suggest using an encrypted session (tpm2_startauthsession), using a trusted primary key.

Unfortunately my dev box is down at the moment. But I think it will look something like:
```bash
tpm2_createprimary -c primry.ctx
# IMPORTANT: verify name from output here
tpm2_startauthsession -c primary.ctx -S session.ctx
echo "my secret" | tpm2_create -S session.ctx -C primary.ctx -u seal.pub -r seal.priv -i-
tpm2_load -S session.ctx -C primary.ctx -u seal.pub -r seal.priv -c seal.ctx
tpm2_unseal -S session.ctx -c seal.ctx
tpm2_flushcontext session.ctx
```
You can add any passwords or policies to objects as you see fit.

**Comment by idesai** [comment]
> Weird that the manpage shows us manually setting attributes like that. Try dropping the `-a` argument to tpm2_create. The reason it's failing is that non-keyedhash objects, in this case rsa, need to have both encrypt and/or decrypt attribute set.
> 
> @idesai is the manpage wrong, should we just drop the `-a` option to `tpm2_create` here?

The -a is needed but with the right set of attributes. Also the man page isn't sufficiently clear that they are two separate examples. I will add the fixes. @x-logic-iot thank you bringing this up.

**Comment by x-logic-iot** [comment]
@idesai no problem.

@williamcroberts I tried your commands and they worked well, but without session.ctx files.

I tried it like this:

```
tpm2_createprimary -c primary.ctx
echo "my secret" | tpm2_create -C primary.ctx -u seal.pub -r seal.priv -i-
tpm2_load -C primary.ctx -u seal.pub -r seal.priv -c seal.ctx
tpm2_unseal -c seal.ctx
```

This works well, now the only thing missing is storing "my secret" to the persistent TPM memory. "my secret" should be stored in TPM and never leave TPM unencrypted. Is there any way of doing this?

I tried it with this commands:

```
tpm2_evictcontrol -c seal.ctx 0x81010001
tpm2_unseal -c 0x81010001
```

Now the only problem is that with `tpm2_unseal -c 0x81010001` anybody can see "my secret" in plain-text.
